### ISSUE TITLE
Read support for remaining LLVM IR language concepts

### ISSUE THREAD

**Issue by mewmew** [llir/llvm#15]
The intention is to provide read support for LLVM IR assembly using a Gocc generated lexer and parser from a BNF grammar of the LLVM IR assembly language.

The BNF grammar is located at [ast/internal/ll.bnf](https://github.com/llir/llvm/blob/master/asm/internal/ll.bnf). The reason to keep the grammar in an `internal` directory, is because the lexer and parser packages generated by Gocc will be considered internal packages, and should not be used by end-users directly. Instead, high-level libraries will make use of these internal packages to parse LLVM IR assembly into the data structures of the [llir/llvm/ir](https://godoc.org/github.com/llir/llvm/ir) package.

Since LLVM IR makes use of unnamed local variables and basic blocks, a context is required to keep track of and map local IDs to their associated values. A bit unfortunate, but this essentially means we cannot use syntax directed translation to translate directly from LLVM IR assembly to the data structures of the [ir](https://godoc.org/github.com/llir/llvm/ir) package. Instead, we must introduce an intermediate step which keeps the necessary information around for us to create and make use of this contextual information. Said and done, the current approach is to define an [ast](https://godoc.org/github.com/llir/llvm/asm/ast) package for LLVM IR assembly, which will later be traversed to create the aforementioned context and translate AST nodes into their corresponding [ir](https://godoc.org/github.com/llir/llvm/ir) data types.

To get a feel for what the production action expressions of Gocc looks like, see the follow example.

``` go
FuncDef
    : "define" OptFuncLinkage
      FuncHeader FuncBody                         << irx.NewFuncDef($2, $3) >>
;
```
## Help wanted

If anyone manages to figure out a clean way for us to skip this step (i.e. not having to translate from [BNF grammar](https://github.com/llir/llvm/blob/master/asm/internal/ll.bnf) to AST, then from AST to [ir](https://godoc.org/github.com/llir/llvm/ir) data types; but instead, translating directly from [BNF grammar](https://github.com/llir/llvm/blob/master/asm/internal/ll.bnf) to [ir](https://godoc.org/github.com/llir/llvm/ir) data types), and go directly from the [BNF grammar](https://github.com/llir/llvm/blob/master/asm/internal/ll.bnf) to the [ir](https://godoc.org/github.com/llir/llvm/ir) package data types using production action expressions, please let us know. This would facilitate the maintainability and future development of this package a lot!

**Comment by mewmew** [comment]
@quarnster Do you have any idea how we may get rid of the intermediate step of translating into and out of the AST representation? I really wish we could figure how a way, as this extra step will make the threshold higher for keeping in sync with updates to the LLVM IR assmebly language, as time goes by and LLVM develops upstream.

**Comment by mewmew** [comment]
Tracking of read support for individual LLVM IR instructions.

Last update: *2016-09-06*

# Instructions

## Binary Operations

- [ ] Add
- [ ] FAdd
- [ ] Sub
- [ ] FSub
- [ ] Mul
- [ ] FMul
- [ ] UDiv
- [ ] SDiv
- [ ] FDiv
- [ ] URem
- [ ] SRem
- [ ] FRem

## Bitwise Binary Operations

- [ ] ShL
- [ ] LShR
- [ ] AShR
- [ ] And
- [ ] Or
- [ ] Xor

## Vector Operations

- [ ] ExtractElement
- [ ] InsertElement
- [ ] ShuffleVector

## Aggregate Operations

- [ ] ExtractValue
- [ ] InsertValue

## Memory Access and Addressing Operations

- [ ] Alloca
- [ ] Load
- [ ] CmpXchg
- [ ] AtomicRMW
- [ ] GetElementPtr

## Conversion Operations

- [ ] Trunc
- [ ] ZExt
- [ ] SExt
- [ ] FPTrunc
- [ ] FPExt
- [ ] FPToUI
- [ ] FPToSI
- [ ] UIToFP
- [ ] SIToFP
- [ ] PtrToInt
- [ ] IntToPtr
- [ ] BitCast
- [ ] AddrSpaceCast

## Other Operations

- [ ] ICmp
- [ ] FCmp
- [ ] PHI
- [ ] Select
- [ ] Call
- [ ] VAArg
- [ ] LandingPad

# Terminators

- [ ] Ret
- [ ] Jmp
- [ ] Br
- [ ] Switch
- [ ] IndirectBr
- [ ] Invoke
- [ ] Resume
- [ ] Unreachable

**Comment by sangisos** [comment]
@mewmew Is there anything specific I can try to help with?

**Comment by mewmew** [comment]
> @mewmew Is there anything specific I can try to help with?

While not a contrived example, a94bf85bd46557d821ec776320b2603a090b3e46 implemented support for fixing the dummy values of `icmp`, `phi`, `select` and `call` instructions. It may be used as a template for implementing support for the remaining instructions.

Ping me before starting any larger work, so we don't duplicate our efforts. The good thing is, we can easily split the work based on the categories of instructions. I've recently been working on the [other operators](http://llvm.org/docs/LangRef.html#otherops) category.

**Comment by chrisbdaemon** [comment]
Is this particular topic still under active development? Is there any kind of an ETA on when it will be useable?

**Comment by mewmew** [comment]
> Is this particular topic still under active development? Is there any kind of an ETA on when it will be useable?

Hi @chrisbdaemon,

It is very much indeed. Check the [parser branch](https://github.com/llir/llvm/tree/parser) for the latest development.

# Current status

*Status as of 2016-11-29*

Note, the most difficult instructions (i.e. `call`, `br`) have been implemented. Trivial instructions will be generated later on, which includes all instructions in the *binary instructions*, *bitwise instructions* and *conversion instructions* categories.

This leaves primarily the `phi` instruction which we shall try to implement tomorrow.

If all goes well, the parser branch should be read for merge with the master branch within the next few weeks : )

## Instructions

### Binary instructions

- [x] Add
- [ ] FAdd
- [ ] Sub
- [ ] FSub
- [x] Mul
- [ ] FMul
- [ ] UDiv
- [ ] SDiv
- [ ] FDiv
- [ ] URem
- [ ] SRem
- [ ] FRem

### Bitwise instructions

- [ ] Shl
- [ ] LShr
- [ ] AShr
- [ ] And
- [ ] Or
- [ ] Xor

### Memory instructions

- [ ] Alloca
- [x] Load
- [x] Store
- [ ] GetElementPtr

### Conversion instructions

- [ ] Trunc
- [ ] ZExt
- [ ] SExt
- [ ] FPTrunc
- [ ] FPExt
- [ ] FPToUI
- [ ] FPToSI
- [ ] UIToFP
- [ ] SIToFP
- [ ] PtrToInt
- [ ] IntToPtr
- [ ] BitCast
- [ ] AddrSpaceCast

### Other instructions

- [x] ICmp
- [ ] FCmp
- [ ] Phi
- [ ] Select
- [x] Call

## Terminators

- [x] Ret
- [x] Br
- [x] CondBr
- [ ] Switch
- [ ] Unreachable

**Comment by mewmew** [comment]
Before the v0.2 release, we should try to implement support for the subset of LLVM IR produced by clang for regular C programs. In particular, anything that we currently [strip](https://github.com/llir/llvm/blob/parser/asm/internal/testdata/strip.sh) before parsing, should be implemented natively by the BNF specification.

This includes
- [ ] `source_filename` specifiers
- [ ] `target` specifiers
- [ ] function `attributes`
- [ ] function attribute IDs (i.e. `#[0-9]+`)
- [ ] metadata nodes and declarations
- [ ] `nsw`, `nuw`, `inbounds`, `signext`
- [ ] linkage (e.g. `common`)
- [ ] ...

Help fill out this list as you encounter more. Our goal is to remove the need for `strip.sh`.

Note, the intention is to provide support for parsing source files including these directives, not to define the data types holding information about these directives, for the v0.2 release of the LLVM IR library.

Future releases of the LLVM IR library will track the development of a sane data representation for these directives. For now, we simply want to be able to parse *complex* LLVM IR files, but only want to *retain* the most relevant information.

**Comment by mewmew** [comment]
Add support for unnamed function parameters. Example:

```llvm
; Function Attrs: nounwind uwtable
define i32 @add(i32, i32) #0 {
  %3 = alloca i32, align 4
  %4 = alloca i32, align 4
  store i32 %0, i32* %3, align 4
  store i32 %1, i32* %4, align 4
  %5 = load i32, i32* %3, align 4
  %6 = load i32, i32* %4, align 4
  %7 = add nsw i32 %5, %6
  ret i32 %7
}
```

**Comment by mewmew** [comment]
The LLVM IR library read support is moving closer to being usable for real world LLVM IR assembly files.

For now, we've been playing with adding read support for c4.ll and [sqlite3.ll](https://github.com/llir/llvm/blob/master/asm/internal/testdata/sqlite/sqlite3.ll), produced by Clang from [c4.c](https://github.com/rswier/c4/blob/master/c4.c) and [sqlite3.c](https://github.com/llir/llvm/blob/master/asm/internal/testdata/sqlite/sqlite3.c), respectively.

The Gocc generated parser is now very close at handling these files.

For c4.ll when optimized with `opt --mem2reg`, support for `undef` values are not yet added. Comments on how to do so would be appreciated. There have been a proposal to potentially remove `undef` from LLVM IR. I still haven't distilled it, so have no gut feeling if I agree with the proposed solution, but still wish to follow its progress; http://llvm.org/devmtg/2016-11/#talk13

For sqlite3.ll a single instruction remains to be implemented, namely [fence](http://llvm.org/docs/LangRef.html#fence-instruction).

It is exciting times!!

This work is tracked in tandem with the development of an LLVM IR to Go decompiler, which uses these files as a first [semi-real world test case](https://github.com/decomp/decomp/issues/185) : )

**Comment by mewmew** [comment]
Left to do as of 2017-04-02:

* [ ] [Other Values](http://llvm.org/docs/LangRef.html#other-values)
* [ ] [Metadata Values](http://llvm.org/docs/LangRef.html#metadata)

**Comment by mewmew** [comment]
@chrisbdaemon 

> Is this particular topic still under active development? Is there any kind of an ETA on when it will be useable?

We are currently preparing for the v0.2 release of llir/llvm. Within the next few days we will check-in the source code of the Gocc generated parser into the source tree, thus making the project go-getable.

https://github.com/llir/llvm/issues/15#issuecomment-290955212 tracks the last few pieces of the LLVM IR assembly language which are currently not part of the BNF grammar. A few of those language concepts have been intentionally postponed to a future release, as their occurrence in LLVM IR source files are not too common.

Cheers /u & i

**Comment by mewmew** [comment]
@chrisbdaemon The source code of the Gocc generated parser has now been checked into the source tree, thus making the llir/llvm packages go-getable.

The remaining LLVM IR language concepts are tracked for the version 0.3 release (https://github.com/llir/llvm/issues/15#issuecomment-290955212).

**Comment by mewmew** [comment]
Read and write support for all LLVM IR construct as of LLVM 7.0 have now been implemented. Therefore, we may now close this issue. If you find any constructs that are not supported, feel free to comment here or open a new issue.
